bindings/node/test/autoEncrypter.test.js

```diff
@@ -0,0 +1,153 @@
+);
+const MOCK_KEYDOCUMENT_RESPONSE = readExtendedJsonToBuffer(`${__dirname}/data/key-document.json`);
+const MOCK_KMS_DECRYPT_REPLY = readHttpResponse(`${__dirname}/data/kms-decrypt-reply.txt`);
+
+class MockClient {
+  constructor() {
+    this.topology = {
+      bson: BSON
+    };
+  }
+}
+
+const AutoEncrypter = require('../lib/autoEncrypter').AutoEncrypter;
+describe('AutoEncrypter', function() {
+  let ENABLE_LOG_TEST = false;
+  let sandbox = sinon.createSandbox();
+  before(() => {
+    sandbox.stub(StateMachine.prototype, 'kmsRequest').callsFake(request => {
+      request.addResponse(MOCK_KMS_DECRYPT_REPLY);
+      return Promise.resolve();
+    });
+
+    sandbox
+      .stub(StateMachine.prototype, 'fetchCollectionInfo')
+      .callsFake((client, ns, filter, callback) => {
+        callback(null, MOCK_COLLINFO_RESPONSE);
+      });
+
+    sandbox
+      .stub(StateMachine.prototype, 'markCommand')
+      .callsFake((client, ns, command, callback) => {
+        if (ENABLE_LOG_TEST) {
+          const response = BSON.deserialize(MOCK_MONGOCRYPTD_RESPONSE);
+          response.schemaRequiresEncryption = false;
+
+          ENABLE_LOG_TEST = false; // disable test after run
+          callback(null, BSON.serialize(response));
+          return;
+        }
+
+        callback(null, MOCK_MONGOCRYPTD_RESPONSE);
+      });
+
+    sandbox.stub(StateMachine.prototype, 'fetchKeys').callsFake((client, ns, filter, callback) => {
+      // mock data is already seriaized, our action deals with the result of a cursor
+      const deserializedKey = BSON.deserialize(MOCK_KEYDOCUMENT_RESPONSE);
+      callback(null, [deserializedKey]);
+    });
+  });
+
+  after(() => {
+    sandbox.restore();
+  });
+
+  describe('state machine', function() {
+    it('should decrypt mock data', function(done) {
+      const input = readExtendedJsonToBuffer(`${__dirname}/data/encrypted-document.json`);
+      const client = new MockClient();
+      const mc = new AutoEncrypter(client, {
+        keyVaultNamespace: 'admin.datakeys',
+        logger: () => {},
+        kmsProviders: {
+          aws: { accessKeyId: 'example', secretAccessKey: 'example' },
+          local: { key: Buffer.alloc(96) }
+        }
+      });
+
+      mc.decrypt(input, (err, decrypted) => {
+        if (err) return done(err);
+        expect(decrypted).to.eql({ filter: { find: 'test', ssn: '457-55-5462' } });
+        done();
+      });
+    });
+
+    it('should encrypt mock data', function(done) {
+      const client = new MockClient();
+      const mc = new AutoEncrypter(client, {
+        keyVaultNamespace: 'admin.datakeys',
+        logger: () => {},
+        kmsProviders: {
+          aws: { accessKeyId: 'example', secretAccessKey: 'example' },
+          local: { key: Buffer.alloc(96) }
+        }
+      });
+
+      mc.encrypt('test.test', TEST_COMMAND, (err, encrypted) => {
+        if (err) return done(err);
+        expect(encrypted).to.have.keys('filter', 'find');
```

Comment:
I'm also confused by this test. Do we assert that the encryption was successful?

Reply:
so, we can't really assert the encrypted values, so what we are doing is asserting that the keys still exist

Reply:
Why can't we assert the encrypted values?

Reply:
updated to assert the deterministically encrypted value